### ISSUE TITLE
confuse on implementation of DoE module

### ISSUE THREAD

**Issue by xxchauncey** [karlstratos/doe#1]
Hi, I'm really confused about the code here: https://github.com/karlstratos/doe/blob/master/util.py#L22, why the returned should be detached like (mi_loss - loss).detach() + loss instead of mi_loss directly?

is it a trick to stabilize the training process? If so, what will happen if i directly return mi_loss?

**Comment by link-er** [comment]
I have the same issue. It seems to be used in several implementations of similar estimators (https://github.com/ermongroup/smile-mi-estimator/blob/master/estimators.py#L123), but I cannot find any explanations in the paper

**Comment by karlstratos** [comment]
Hi @xxchauncey, @link-er, sorry about the late reply! This is just a computation graph trick to separate the training loss from the return value. It allows us to collect the MI estimate like [this](https://github.com/karlstratos/doe/blob/84b5d5c394d5e6be3455a48ced742c231e2fef5c/main.py#L95) uniformly for all models (instead of having to recalculate estimator-specific values). It's a well-known trick useful for things like straight-through estimators and gradient reversal, but it's used just for convenience here. Does this make sense?

**Comment by xxchauncey** [comment]
oh, I see! smart trick!
thank you for your reply!

**Comment by xxchauncey** [comment]
> Hi @xxchauncey, @link-er, sorry about the late reply! This is just a computation graph trick to separate the training loss from the return value. It allows us to collect the MI estimate like [this](https://github.com/karlstratos/doe/blob/84b5d5c394d5e6be3455a48ced742c231e2fef5c/main.py#L95) uniformly for all models (instead of having to recalculate estimator-specific values). It's a well-known trick useful for things like straight-through estimators and gradient reversal, but it's used just for convenience here. Does this make sense?

oh, I see! smart trick!
thank you for your reply!
